sklearn/model_selection/_classification_threshold.py

```diff
@@ -0,0 +1,924 @@
+
+    This estimator post-tunes the decision threshold (cut-off point) that is
+    used for converting posterior probability estimates (i.e. output of
+    `predict_proba`) or decision scores (i.e. output of `decision_function`)
+    into a class label. The tuning is done by maximizing a binary metric,
+    potentially constrained by a another metric.
+
+    Read more in the :ref:`User Guide <tunedthresholdclassifier>`.
+
+    .. versionadded:: 1.5
+
+    Parameters
+    ----------
+    estimator : estimator instance
+        The classifier, fitted or not, for which we want to optimize
+        the decision threshold used during `predict`.
+
+    strategy : {"optimum", "constant"}, default="optimum"
+        The strategy to use for tuning the decision threshold:
+
+        * `"optimum"`: the decision threshold is tuned to optimize the objective
+            metric;
+        * `"constant"`: the decision threshold is set to `constant_value`.
+
+    objective_metric : {"max_tpr_at_tnr_constraint", "max_tnr_at_tpr_constraint", \
+            "max_precision_at_recall_constraint, "max_recall_at_precision_constraint"} \
+            , str, dict or callable, default="balanced_accuracy"
+        The objective metric to be optimized. Can be one of:
+
+        * a string associated to a scoring function (see model evaluation
+          documentation);
+        * a scorer callable object created with :func:`~sklearn.metrics.make_scorer`;
+        * `"max_tnr_at_tpr_constraint"`: find the decision threshold for a true
+          positive ratio (TPR) of `constraint_value`;
+        * `"max_tpr_at_tnr_constraint"`: find the decision threshold for a true
+          negative ratio (TNR) of `constraint_value`.
+        * `"max_precision_at_recall_constraint"`: find the decision threshold for a
+          recall of `constraint_value`;
+        * `"max_recall_at_precision_constraint"`: find the decision threshold for a
+          precision of `constraint_value`.
+
+    constraint_value : float, default=None
+        The value associated with the `objective_metric` metric for which we
+        want to find the decision threshold when `objective_metric` is either
+        `"max_tnr_at_tpr_constraint"`, `"max_tpr_at_tnr_constraint"`,
+        `"max_precision_at_recall_constraint"`, or
+        `"max_recall_at_precision_constraint"`.
+
+    constant_threshold : float, default=0.5
+        The constant threshold to use when `strategy` is `"constant"`.
+
+    pos_label : int, float, bool or str, default=None
+        The label of the positive class. Used when `objective_metric` is
+        `"max_tnr_at_tpr_constraint"`"`, `"max_tpr_at_tnr_constraint"`, or a dictionary.
+        When `pos_label=None`, if `y_true` is in `{-1, 1}` or `{0, 1}`,
+        `pos_label` is set to 1, otherwise an error will be raised. When using a
+        scorer, `pos_label` can be passed as a keyword argument to
+        :func:`~sklearn.metrics.make_scorer`.
```

Comment:
Instead of thinking of our internal code, I think that we need to think about the potential use case where `pos_label` will be involved.

A user will have to use `pos_label` when defining a scorer if they want to evaluate a model, independently if they use `TunedThresholdClassifierCV`. So somehow, we could expect that the user is creating a scorer and pass `pos_label`.

The tricky part is that we need to have `pos_label` in `TunedThresholdClassifierCV` to forward it to the scorer when the this is a constraint scorer because we have no way to do otherwise.

So if we want to support both cases, it seems more convenient to also accept `pos_label` for normal metric and validate to be sure that we don't have an inconsistency and thus the weird double validation.

Alternatively, we could raise an error if you pass a scorer and `pos_label` instead of making the difficult validation but it could be potentially more annoying using it.

Reply:
Nevertheless, I think we should validate `pos_label`, once, right away. We have access to the `pos_label` parameter and to the scorer at the beginning of fit.